sdk/iOS/OoyalaSkinSDK/ViewControllers/FullscreenStateController/FullscreenStateController.m

```diff
@@ -0,0 +1,175 @@
+                     containerView:(UIView *)containerView
+                         videoView:(UIView *)videoView
+       andFullscreenViewController:(UIViewController *)fullscreenViewController {
+  self = [super init];
+  if (self) {
+    self.parentView = parentView;
+    self.containerView = containerView;
+    self.videoView = videoView;
+    self.fullscreenViewController = fullscreenViewController;
+    
+    [self configure];
+  }
+  return self;
+}
+
+#pragma mark - Public functions
+
+- (void)setFullscreen:(BOOL)fullscreen completion:(nullable void (^)())completion {
+  for (NSOperation *operation in self.operationQueue.operations) {
+    if (!operation.isExecuting) {
+      [operation cancel];
+    }
+  }
+  
+  NSOperation *operation = [[FullscreenStateOperation alloc] initWithFullscreen:fullscreen enterFullscreenBlock:^(void (^animationCompletion)(void)) {
+    if (self.isFullscreen == fullscreen) {
+      animationCompletion();
+    } else {
+      [self openFullscreenMode:^{
+        animationCompletion();
+      }];
+    }
+  } enterInlineBlock:^(void (^animationCompletion)(void)) {
+    if (self.isFullscreen != fullscreen) {
+      [self openInlineMode:^{
+        animationCompletion();
+      }];
+    } else {
+      animationCompletion();
+    }
+  } andCompleteStateChanges:completion];
+  
+  [self.operationQueue addOperation:operation];
+}
+
+#pragma mark - Private functions
+
+- (void)configure {
+  self.operationQueue = [NSOperationQueue new];
+  [self.operationQueue setMaxConcurrentOperationCount:1];
+}
+
+- (void)openFullscreenMode:(nullable void (^)())completion {
+  UIWindow *window = [UIApplication sharedApplication].keyWindow;
+  
+  // Save root VC
+  
+  self.rootViewController = window.rootViewController;
+  
+  // Remove video view from container
+  
+  [self.videoView removeFromSuperview];
+  
+  // Configure fullscreen VC
+  
+  self.fullscreenViewController.view.frame = window.bounds;
+  [self.fullscreenViewController.view addSubview:self.videoView];
+  
+  // Convert rect for animation
+  
+  CGRect frameInWindow = [self.parentView convertRect:self.containerView.frame
+                                               toView:window];
+  
+  // Set start rect for video view
+  
+  self.videoView.frame = frameInWindow;
+  
+  // Add fullscreen VC on window as subview
+  
+  [window addSubview:self.fullscreenViewController.view];
+  [window bringSubviewToFront:self.fullscreenViewController.view];
+  
+  // Perform animation
+  
+  [UIView animateWithDuration:FULLSCREEN_ANIMATION_DURATION animations:^{
+    self.videoView.frame = window.bounds;
+  } completion:^(BOOL finished) {
+    
+    // Change root VC
+    
+    window.rootViewController = self.fullscreenViewController;
+    
+    // Update current fullscreen state
+    
+    self.isFullscreen = YES;
+    
+    // Completion
+    
+    if (completion) {
+      completion();
+    }
+  }];
+}
+
+- (void)openInlineMode:(nullable void (^)())completion {
```

Comment:
We need docs for all these methods. I don't understand what this one does.

Reply:
I created s separate tickets that will be included into next sprint to write docs for new methods - https://jira.corp.ooyala.com/browse/PLAYER-3138